compiler/rustc_typeck/src/coherence/inherent_impls.rs

```diff
@@ -124,14 +119,52 @@ impl<'tcx> ItemLikeVisitor<'_> for InherentCollect<'tcx> {
     fn visit_foreign_item(&mut self, _foreign_item: &hir::ForeignItem<'_>) {}
 }
 
+const INTO_CORE: &str = "consider moving this inherent impl into `core` if possible";
+const INTO_DEFINING_CRATE: &str =
+    "consider moving this inherent impl into the crate defining the type if possible";
+const ADD_ATTR: &str =
+    "alternatively add `#[rustc_allow_incoherent_impl]` to the relevant impl items";
+
 impl<'tcx> InherentCollect<'tcx> {
-    fn check_def_id(&mut self, item: &hir::Item<'_>, def_id: DefId) {
+    fn check_def_id(&mut self, item: &hir::Item<'_>, self_ty: Ty<'tcx>, def_id: DefId) {
+        let impl_def_id = item.def_id;
         if let Some(def_id) = def_id.as_local() {
             // Add the implementation to the mapping from implementation to base
             // type def ID, if there is a base type for this implementation and
             // the implementation does not have any associated traits.
             let vec = self.impls_map.inherent_impls.entry(def_id).or_default();
-            vec.push(item.def_id.to_def_id());
+            vec.push(impl_def_id.to_def_id());
+            return;
+        }
+
+        if self.tcx.features().rustc_attrs
```

Comment:
not completely, `self.tcx.features().rustc_attrs` is about the current crate while `self.tcx.has_attr(def_id, sym::rustc_has_incoherent_inherent_impls)` is for the crate defining the type. Without the `rustc_attrs` condition we would allow empty impls for foreign types on stable.

Reply:
Yep, you are right about the empty impls.

But then the condition should be
```rust
if assoc_items.is_empy() || !assoc_items.all(|item| item.has_attr(rustc_has_incoherent_inherent_impls)) {
    report_error();
}
```
, the `self.tcx.features().rustc_attrs` condition is still irrelevant.